src/iceberg/parquet/parquet_schema_util.cc

```diff
@@ -17,20 +17,392 @@
+        }
+      }
+      break;
+    case TypeId::kStruct:
+      if (arrow_type->id() == ::arrow::Type::STRUCT) {
+        return {};
+      }
+      break;
+    case TypeId::kList:
+      if (arrow_type->id() == ::arrow::Type::LIST) {
+        return {};
+      }
+      break;
+    case TypeId::kMap:
+      if (arrow_type->id() == ::arrow::Type::MAP) {
+        return {};
+      }
+      break;
+    default:
+      break;
+  }
+
+  return InvalidSchema("Cannot read Iceberg type: {} from Parquet type: {}",
+                       expected_type, arrow_type->ToString());
+}
+
+// Forward declaration
+Result<FieldProjection> ProjectNested(
+    const Type& expected_type,
+    const std::vector<::parquet::arrow::SchemaField>& parquet_fields);
+
+Result<FieldProjection> ProjectStruct(
+    const StructType& struct_type,
+    const std::vector<::parquet::arrow::SchemaField>& parquet_fields) {
+  struct FieldContext {
+    size_t local_index;
+    const ::parquet::arrow::SchemaField& parquet_field;
+  };
+  std::unordered_map<int32_t, FieldContext> field_context_map;
+  field_context_map.reserve(parquet_fields.size());
+
+  for (size_t i = 0; i < parquet_fields.size(); ++i) {
+    const ::parquet::arrow::SchemaField& parquet_field = parquet_fields[i];
+    auto field_id = GetFieldId(parquet_field);
+    if (!field_id) {
+      continue;
+    }
+    if (const auto [iter, inserted] = field_context_map.emplace(
+            std::piecewise_construct, std::forward_as_tuple(field_id.value()),
+            std::forward_as_tuple(i, parquet_field));
+        !inserted) [[unlikely]] {
+      return InvalidSchema("Duplicate field id found in Parquet schema: {}",
+                           field_id.value());
+    }
+  }
+
+  FieldProjection result;
+  result.children.reserve(struct_type.fields().size());
+
+  for (const auto& expected_field : struct_type.fields()) {
+    int32_t field_id = expected_field.field_id();
+    FieldProjection child_projection;
+
+    if (auto iter = field_context_map.find(field_id); iter != field_context_map.cend()) {
+      const auto& parquet_field = iter->second.parquet_field;
+      ICEBERG_RETURN_UNEXPECTED(
+          ValidateParquetSchemaEvolution(*expected_field.type(), parquet_field));
+      if (expected_field.type()->is_nested()) {
+        ICEBERG_ASSIGN_OR_RAISE(child_projection, ProjectNested(*expected_field.type(),
+                                                                parquet_field.children));
+      } else {
+        child_projection.attributes =
+            std::make_shared<ParquetExtraAttributes>(parquet_field.column_index);
+      }
+      child_projection.from = iter->second.local_index;
+      child_projection.kind = FieldProjection::Kind::kProjected;
+    } else if (MetadataColumns::IsMetadataColumn(field_id)) {
+      child_projection.kind = FieldProjection::Kind::kMetadata;
+    } else if (expected_field.optional()) {
+      child_projection.kind = FieldProjection::Kind::kNull;
```

Comment:
(So only struct can contain these `kMetadata` and `kNull` projections. List entry and map value cannot?